tests/compiler/test_list_comprehension.py

```diff
@@ -290,49 +290,136 @@ def test_list_comprehension_nested_tail(snippetcompiler) -> None:
     compiler.do_compile()
 
 
+# TODO: tests failures for new tests on master
 def test_list_comprehension_gradual(snippetcompiler) -> None:
     """
     Verify that list comprehensions are executed gradually.
     """
     snippetcompiler.setup_for_snippet(
         textwrap.dedent(
             """
-            entity A: end
+            entity A:
+                string? name = null
+            end
+            A.opt [0:1] -- A
             A.others [0:] -- A
             implement A using std::none
 
             # gradual execution of iterable
-            a = A()
-            b = A(others=[chained for chained in [other for other in a.others]])
+            a = A(name="a")
+            b = A(name="b", others=[chained for chained in [other for other in a.others]])
             a.others += A()
             if b.others is defined:
                 # this seems like bad practice, but it should work as long as the list comprehension is executed gradually.
                 a.others += A()
             end
 
             # gradual execution of guard
-            x = A()
+            x = A(name="x")
             # again bad practice but asserts gradual execution of the `is defined`
-            y = A(others=[A(), c.others])
-            c = A(others=[candidate for candidate in [x, y] if candidate.others is defined])
+            y = A(name="y", others=[A(), c.others])
+            c = A(name="c", others=[candidate for candidate in [x, y] if candidate.others is defined])
 
             # gradual execution with nested lists in the iterable
-            u = A()
-            v = A(others=[candidate for candidate in [a, [b, c, [x, y, u], u, x], y]])
+            u = A(name="u")
+            v = A(name="v", others=[candidate for candidate in [a, [b, c, [x, y, u], u, x], y]])
             if v.others is defined:
                 # again bad practice but asserts gradual execution of the `is defined`
                 v.others += A()
             end
 
-            count = 2
-            count = std::count(a.others)
-            count = std::count(b.others)
+            # gradual execution with nested for
+            w = A(
+                name="w",
+                others=[
+                    o
+                    for base in v.others
+                    for o in base.others
+                ]
+            )
+            if w.others is defined:
+                # again bad practice but asserts gradual execution of the `is defined`
+                w.others += A()
+            end
+
+            # verify that a constructor expression does not block gradual execution,
+            # which would result in a list modified after freeze
```

Comment:
This is actually the only one that was broken on master. The others were inspired by issues / non-trivial edge cases I encountered during the refactor.